File: src/brs/Generator.java
package brs;

import brs.util.Observable;
import brs.util.ThreadPool;

import java.math.BigInteger;
import java.util.Collection;

public interface Generator extends Observable<Generator.GeneratorState, Generator.Event> {

  enum Event {
    GENERATION_DEADLINE, NONCE_SUBMITTED
  }

  void generateForBlockchainProcessor(ThreadPool threadPool, BlockchainProcessor blockchainProcessor);

  GeneratorState addNonce(String secretPhrase, Long nonce);

  GeneratorState addNonce(String secretPhrase, Long nonce, byte[] publicKey);

  Collection<GeneratorState> getAllGenerators();

  byte[] calculateGenerationSignature(byte[] lastGenSig, long lastGenId);

  int calculateScoop(byte[] genSig, long height);

  BigInteger calculateHit(long accountId, long nonce, byte[] genSig, int scoop, int blockHeight);

  BigInteger calculateHit(long accountId, long nonce, byte[] genSig, byte[] scoopData);

  BigInteger calculateDeadline(long accountId, long nonce, byte[] genSig, int scoop, long baseTarget, int blockHeight);

  interface GeneratorState {
    byte[] getPublicKey();

    Long getAccountId();

    BigInteger getDeadline();

    long getBlock();
  }
}

File: src/main/java/brs/Generator.kt
package brs

import brs.util.Observable

import java.math.BigInteger

interface Generator : Observable<Generator.GeneratorState, Generator.Event> {
    val allGenerators: Collection<GeneratorState>

    enum class Event {
        GENERATION_DEADLINE, NONCE_SUBMITTED
    }

    fun generateForBlockchainProcessor(dp: DependencyProvider)

    fun addNonce(secretPhrase: String, nonce: Long?): GeneratorState

    fun addNonce(secretPhrase: String, nonce: Long?, publicKey: ByteArray): GeneratorState

    fun calculateGenerationSignature(lastGenSig: ByteArray, lastGenId: Long): ByteArray

    fun calculateScoop(genSig: ByteArray, height: Long): Int

    fun calculateHit(accountId: Long, nonce: Long, genSig: ByteArray, scoop: Int, blockHeight: Int): BigInteger

    fun calculateHit(accountId: Long, nonce: Long, genSig: ByteArray, scoopData: ByteArray): BigInteger

    fun calculateDeadline(accountId: Long, nonce: Long, genSig: ByteArray, scoop: Int, baseTarget: Long, blockHeight: Int): BigInteger

    interface GeneratorState {
        val publicKey: ByteArray

        val accountId: Long?

        val deadline: BigInteger

        val block: Long
    }
}
